site/docs/tutorial/sql_to_substrait.md

```diff
@@ -0,0 +1,1305 @@
+                                  "nullability": "NULLABILITY_REQUIRED"
+                                }
+                              },
+                              "input": {
+                                "selection": {
+                                  "directReference": {
+                                    "structField": {
+                                      "field": 1
+                                    }
+                                  }
+                                }
+                              }
+                            }
+                          }
+                        },
+                        {
+                          "value": {
+                            "selection": {
+                              "directReference": {
+                                "structField": {
+                                  "field": 3
+                                }
+                              }
+                            }
+                          }
+                        }
+                      ]
+                    }
+                  }
+                }
+              }
+            ]
+          }
+        }
+      }
+    ]
+  }
+}
+```
+
+## Field indices
+
+So far, we have glossed over the field indices. Now that we've built up each
+of the relations, it will be a bit easier to explain them.
+
+Instead of referencing fields by name, in Substrait we always reference them by
+index. Nested fields are part of the index sequence, put in depth-first order.
+For example, as shown in the diagram below, in the `products` table read relation
+the `manufacturer` and `year_created` fields come right after their parent field,
+`details`.
+
+One complication is that whenever we reference by field index, those field
+indices are dependent on the context of the relation. The exact field indices
+can change from relation to relation. For example, a join relations field 
+indices are based on the combination of the left inputs fields and then the 
+right input fields. And, while we don't have one in our example, a Project relation
+adds new fields, which are visible to it's downstream relations. The diagram
+below shoes the mapping of field indices within each relation and how each of
+the field references show up in each relations properties.
+
+![](field_indices_layout.svg)
+
+## Column selection and emit
+
+Looking at the relations so far, you might be wondering: when do we select the
+output columns?
+
+You might be tempted to add a [Project relation](/relations/logical_relations/#project-operation) 
+at the end to select a subset of columns. However, the project relation only
+adds columns; it is not responsible for subsetting or reordering columns.
+
+<!-- TODO: Is this true? -->
+Instead, any relation can reorder or subset columns through the `emit` property.
+By default, it is set to `direct`, which outputs all columns "as is". But it can
+also be specified as a sequence of field indices.
```

Comment:
Just want to make sure? Is this actually true? Is **any** relation able to subset or reorder columns?

Reply:
Yes.  This is quite different from some existing tooling in this area (e.g. Acero, dplyr, ibis) and so if you are familiar with those then it can be a bit odd.  It has been a consistent source of confusion :).

It is a very "logical" concept and allows for writing compact plans.  One way to think about it is to recognize that just about any relation might be the terminal point for a particular column:

 * You might load a column purely for the sake of applying a filter and can drop it after the filter
 * You might load a column purely to use as a join key and drop it after the join
 * Etc.

So, rather than be forced to write a plan that is full of project nodes you can instead just emit at any point to drop a column once you are finished with it.

A corollary confusion is that the project node, by itself, cannot actually drop or reorder columns.  The project node, in Substrait, only adds columns to the end.  The only way to drop columns in a project node is with an emit.

This means that `SELECT expr1, expr2, expr3` typically gets represented as:

`project(expr1, expr2, expr3):emit=[len(input)+0, len(input)+1, len(input)+2]`